chain/client/src/stateless_validation/processing_tracker.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+
+use once_cell::sync::OnceCell;
+
+/// `ProcessingDoneTracker` can be used in conjuction with a `ProcessingDoneWaiter`
+/// to wait until some processing is finished. `ProcessingDoneTracker` should be
+/// kept alive as long as the processing is ongoing, then once it's dropped,
+/// the paired `ProcessingDoneWaiter` will be notified that the processing has finished.\
+/// Does NOT implement `Clone`, if you want to clone it, wrap it in an `Arc`.
+pub struct ProcessingDoneTracker(Arc<OnceCell<()>>);
+
+impl ProcessingDoneTracker {
+    /// Create a new `ProcessingDoneTracker`
+    pub fn new() -> ProcessingDoneTracker {
+        ProcessingDoneTracker(Arc::new(OnceCell::new()))
+    }
+
+    /// Create a `ProcessingDoneWaiter` paired to this `ProcessingDoneTracker`.\
+    /// When this `ProcessingDoneTracker` is dropped, the paired waiter will be notified.
+    pub fn make_waiter(&self) -> ProcessingDoneWaiter {
+        ProcessingDoneWaiter(self.0.clone())
+    }
+}
+
+impl Drop for ProcessingDoneTracker {
+    fn drop(&mut self) {
```

Comment:
Or drop implementation can internally call `done()` as well.

Reply:
Hmm I initialy considered using a `ProcessingDoneTracker::done(self)` method, but I thought that it would be really annoying to call it on all of the code paths. For every `action()?` that there is, we would have to replace it with:
```rust
match action() {
  Ok(x) => x,
  Err(e) => {
    processing_done_tracker.done();
    return Err(e);
   }
}
```
So instead of that I went with a RAII solution, similar to `MutexGuard`. It's pretty intuitive to me, even with all the capturing.

We could have an optional `done()` method, which can be called to make it explicit that we're done, but then there would be two ways to do the same thing, which is a bit of an API smell. A user could be confused - do I have to call `done()` or not? This other code path doesn't call `done()`, why is that?
`MutexGuard` doesn't have an `unlock()` method, so I guess the more rusty way is to use `std::mem::drop` for explicit freeing.

OTOH a `done()` method could be a nice sanity check to make sure that the tracker is kept alive until the very end.

Idk, I'm leaning towards the current version (just `Drop`), but I'm open to alternatives.

Reply:
Yeah, makes sense tbh... I'm fine with the current impl!